### PR TITLE
feat!: implement artifact manifest handler in the storage package

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
Implemented artifact manifest handler related changes for the storage package, unit tests included.

Part 3 of #21 

Signed-off-by: wangxiaoxuan273 <wangxiaoxuan119@gmail.com>

### CODE REVIEW

File: registry/storage/ociartifactmanifesthandler.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	for _, descriptor := range mnfst.References() {
+		err := descriptor.Digest.Validate()
+		if err != nil {
+			errs = append(errs, err, distribution.ErrManifestBlobUnknown{Digest: descriptor.Digest})
+			continue
+		}
+		exists, err := manifestService.Exists(ctx, descriptor.Digest)
+		if err != nil || !exists {
+			errs = append(errs, distribution.ErrManifestBlobUnknown{Digest: descriptor.Digest})
+		}
```
Comment (shizhMSFT) [inline registry/storage/ociartifactmanifesthandler.go:88]: This part does not look right. Descriptor of `Subject` refers to manifest but descriptors of `Blobs` refers to blobs.

File: registry/storage/ociartifactmanifesthandler.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+	manifestService, err := ms.repository.Manifests(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, descriptor := range mnfst.References() {
+		err := descriptor.Digest.Validate()
+		if err != nil {
+			errs = append(errs, err, distribution.ErrManifestBlobUnknown{Digest: descriptor.Digest})
```
Comment (shizhMSFT) [inline registry/storage/ociartifactmanifesthandler.go:82]: `MANIFEST_INVALID` sounds more appropriate as suggested by @m5i-work

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: registry/storage/ociartifactmanifesthandler.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	"fmt"
+
+	"github.com/distribution/distribution/v3"
+	dcontext "github.com/distribution/distribution/v3/context"
+	"github.com/distribution/distribution/v3/manifest/ociartifact"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// ociArtifactManifestHandler is a ManifestHandler that covers ocischema manifests.
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/storage/ociartifactmanifesthandler.go:14]: Nice catch. I forgot to change here.

File: registry/storage/manifeststore.go [hunk 4]
```diff
@@ -121,9 +123,13 @@ func (ms *manifestStore) Get(ctx context.Context, dgst digest.Digest, options ..
 		default:
 			return nil, distribution.ErrManifestVerification{fmt.Errorf("unrecognized manifest content type %s", versioned.MediaType)}
 		}
+	default:
+		// This is bad style but let it be now, it works fine
+		if versioned.MediaType == v1.MediaTypeArtifactManifest {
+			return ms.ociartifactHandler.Unmarshal(ctx, dgst, content)
+		}
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/storage/manifeststore.go:130]: moved.

[the remark this answers is not in the document]
Comment (qweeah) [reply registry/storage/manifeststore.go:130]: Should only check `MediaType` before switching and comment the spec link(https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/artifact.md) in the if block.

File: registry/storage/ociartifactmanifesthandler.go (hunk 1 above)
Comment (wangxiaoxuan273) [reply registry/storage/ociartifactmanifesthandler.go:88]: Corrected this part with `blobService` and `ManifestService`.

File: registry/storage/ociartifactmanifesthandler.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply registry/storage/ociartifactmanifesthandler.go:101]: After offline discussion, we keep using the original errors.

File: registry/storage/manifeststore.go [hunk 5]
```diff
@@ -94,19 +96,25 @@ func (ms *manifestStore) Get(ctx context.Context, dgst digest.Digest, options ..
 		return nil, err
 	}
 
+	switch versioned.MediaType {
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/storage/manifeststore.go:99]: changed accordingly.

File: registry/storage/ociartifactmanifesthandler.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+			exists, err := manifestService.Exists(ctx, descriptor.Digest)
+			if err != nil || !exists {
+				errs = append(errs, distribution.ErrManifestBlobUnknown{Digest: descriptor.Digest})
+			}
+		default:
+			_, err = blobsService.Stat(ctx, descriptor.Digest)
+			if err != nil {
+				errs = append(errs, distribution.ErrManifestBlobUnknown{Digest: descriptor.Digest})
+			}
+		}
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/storage/ociartifactmanifesthandler.go]: Changed the unit test accordingly.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**